File: 0310_Minimum Height Trees/MinimumHeightTrees_BruteForce_TLE.cs
public class Solution {
    public IList<int> FindMinHeightTrees(int n, int[][] edges) {
        var ans = new List<int>();
        if(n < 1) return ans;
        if(n == 1)
        {
            ans.Add(0);
            return ans;
        }
        var graph = BuildGraph(edges);
        var currMHT = int.MaxValue;
        
        for(var node = 0;node<n;node++)
        {
            var depth = Traverse(node, graph, currMHT);
            
            if(depth <= currMHT)
            {
                if(depth < currMHT)
                    ans.Clear();
                
                ans.Add(node);
                currMHT = depth;
            }
        }
        
        return ans;
    }
    
    private Dictionary<int, List<int>> BuildGraph(int[][] edges)
    {
        var graph = new Dictionary<int, List<int>>();
        foreach(var edge in edges)
        {
            if(!graph.ContainsKey(edge[0])) graph.Add(edge[0], new List<int>());
            if(!graph.ContainsKey(edge[1])) graph.Add(edge[1], new List<int>());
            graph[edge[0]].Add(edge[1]);
            graph[edge[1]].Add(edge[0]);
        }
        
        return graph;
    }
    
    private int Traverse(int root, Dictionary<int, List<int>> graph, int currMHT)
    {
        var queue = new Queue<int>();
        queue.Enqueue(root);
        var depth = 0;
        var visited = new HashSet<int>();
        visited.Add(root);
        
        while(queue.Count > 0)
        {
            depth++;
            if(depth > currMHT) return int.MaxValue;
            var c = queue.Count;
            while(c-- > 0)
            {
                var node = queue.Dequeue();
                foreach(var neighbor in graph[node])
                {
                    if(visited.Contains(neighbor)) continue;
                    visited.Add(neighbor);
                    queue.Enqueue(neighbor); 
                }
            }
        }
        
        return depth;
    }
}

File: 0310_Minimum Height Trees/MinimumHeightTrees_BruteForce_TLE.java
class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        List<Integer> ans = new ArrayList<>();
        if(n == 1) {
            ans.add(0);
            return ans;
        }
        
        int minHeight = n;
        Map<Integer, List<Integer>> graph = buildGraph(edges);
        for(int i=0;i<n;i++) {
            int height = height(graph, i);
            if(height == minHeight) {
                ans.add(i);
            }else if(height < minHeight) {
                ans.clear();
                minHeight = height;
                ans.add(i);
            }
        }

        return ans;

    }

    private Map<Integer, List<Integer>> buildGraph(int[][] edges) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for(int[] edge : edges) {
            int p1 = edge[0], p2 = edge[1];
            graph.putIfAbsent(p1, new ArrayList<Integer>());
            graph.putIfAbsent(p2, new ArrayList<Integer>());
            graph.get(p1).add(p2);
            graph.get(p2).add(p1);
        }
        return graph;
    }

    private int height(Map<Integer, List<Integer>> graph, int node) {
        if(!graph.containsKey(node)) return Integer.MAX_VALUE;
        Queue<Integer> q = new LinkedList<>();
        q.offer(node);
        int height = 0;
        Set<Integer> visited = new HashSet<>();
        
        while(q.size() > 0) {
            int size = q.size();
            height++;
            while(size-- > 0) {
                int currNode = q.poll();
                visited.add(currNode);
                for(int nextNode : graph.get(currNode)){
                    if(visited.contains(nextNode)) continue;
                    q.offer(nextNode);
                }
            }
        }

        return height;
    }
}